src/Temporalio/Worker/WorkflowWorker.cs

```diff
@@ -0,0 +1,254 @@
+#pragma warning disable CA1031 // We do want to catch _all_ exceptions in this file sometimes
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Temporalio.Bridge.Api.WorkflowActivation;
+using Temporalio.Bridge.Api.WorkflowCompletion;
+using Temporalio.Exceptions;
+using Temporalio.Workflows;
+
+namespace Temporalio.Worker
+{
+    /// <summary>
+    /// Worker for workflows.
+    /// </summary>
+    internal class WorkflowWorker
+    {
+        private static readonly TimeSpan DefaultDeadlockTimeout = TimeSpan.FromSeconds(2);
+
+        private readonly ILogger logger;
+        // Keyed by run ID
+        private readonly ConcurrentDictionary<string, IWorkflowInstance> runningWorkflows = new();
+        // Keyed by run ID
+        private readonly ConcurrentDictionary<string, Task> deadlockedWorkflows = new();
```

Comment:
You're allowed to not complete a task forever, and the slot will be occupied. Shutdown will never resolve until all outstanding work is replied to.

Reply:
@Sushisource - do you think that's a better approach than what I have today where I do evict but still kinda occupy a slot locally? I'm thinking so because I bet that task I'm holding has a reference to my task scheduler which means the whole workflow is probably still in memory.

Reply:
I think that'd be swell - but you do still want to fail the activation so that the user ends up seeing WFT failure "because deadlocked" in their history. As long as that still happens the approach you describe makes sense to me.

Reply:
Fixed in next PR to not complete the remove-job task until the hung workflow task is complete